Support configuration cache

Newest version of axion-release-plugin doesn't support [configuration cache](https://docs.gradle.org/6.8.3/userguide/configuration_cache.html). Here you can see why it fails:
```sh
./gradlew currentVersion --configuration-cache
2 problems were found storing the configuration cache, 1 of which seems unique.
- Task `:currentVersion` of type `pl.allegro.tech.build.axion.release.OutputCurrentVersionTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/6.8.3/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
```
Both problems are related with using `Project` during task execution:
![image](https://user-images.githubusercontent.com/317662/112760013-c9c7b500-8ff5-11eb-8151-b5cd733257d0.png)
![image](https://user-images.githubusercontent.com/317662/112760049-e7951a00-8ff5-11eb-8dd8-8442d2d5c823.png)

Here you can see how to resolve typical problems with using Project inside a task:
https://docs.gradle.org/6.8.3/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution

It's also related to that issue:
https://github.com/gradle/gradle/issues/13490

Seeing this issue is still opened, I'll share that I started observing the plugin throwing an error when running with Gradle 8.1 (which has the Configuration Cache feature stabilized)

I apply plugin in a convention plugin:
```
class VersioningPlugin : Plugin<Project> {

    override fun apply(target: Project): Unit = with(target) {
        if (this != rootProject) throw GradleException("Versioning plugin can be applied to the root project only")
        pluginManager.apply(ReleasePlugin::class.java)

        val scmConfig = extensions.getByType<VersionConfig>()

        allprojects { project ->
            project.version = scmConfig.version
        }
    }
```
which should somewhat transform into what [Basic Usage](https://github.com/allegro/axion-release-plugin#basic-usage) section presents. 

The error is
```
FAILURE: Build failed with an exception.

* What went wrong:
Configuration cache problems found in this build.

2 problems were found storing the configuration cache.
- Plugin 'com.starter.versioning': external process started '/usr/local/bin/git --version'
  See https://docs.gradle.org/8.1/userguide/configuration_cache.html#config_cache:requirements:external_processes
- Plugin 'com.starter.versioning': external process started '/usr/local/bin/git config --system --show-origin --list -z'
  See https://docs.gradle.org/8.1/userguide/configuration_cache.html#config_cache:requirements:external_processes

See the complete report at file://{project_path}/build/reports/configuration-cache/bw6o1r9owtj2rruhd9167sy1u/663y7tyiq194f3u0xfz72ugjn/configuration-cache-report.html
> Starting an external process '/usr/local/bin/git --version' during configuration time is unsupported.
> Starting an external process '/usr/local/bin/git config --system --show-origin --list -z' during configuration time is unsupported.
```
\+
<details>
<summary>one of stacktraces from `configuration-cache-report.html`</summary>

```stacktrace
org.gradle.api.InvalidUserCodeException: Starting an external process '/usr/local/bin/git --version' during configuration time is unsupported.
	at org.gradle.configurationcache.initialization.DefaultConfigurationCacheProblemsListener.onExternalProcessStarted(ConfigurationCacheProblemsListener.kt:87)
	at org.gradle.configurationcache.InstrumentedInputAccessListener.externalProcessStarted(InstrumentedInputAccessListener.kt:110)
	at org.gradle.internal.classpath.Instrumented.externalProcessStarted(Instrumented.java:501)
	at org.gradle.internal.classpath.Instrumented.externalProcessStarted(Instrumented.java:509)
	at org.gradle.internal.classpath.Instrumented.start(Instrumented.java:363)
	at org.eclipse.jgit.util.FS.readPipe(FS.java:1379)
	at org.eclipse.jgit.util.FS.readPipe(FS.java:1341)
	at org.eclipse.jgit.util.FS.discoverGitSystemConfig(FS.java:1530)
	at org.eclipse.jgit.util.FS.getGitSystemConfig(FS.java:1627)
	at org.eclipse.jgit.util.SystemReader$Default.openSystemConfig(SystemReader.java:86)
	at org.eclipse.jgit.util.SystemReader.getSystemConfig(SystemReader.java:363)
	at org.eclipse.jgit.util.SystemReader.getUserConfig(SystemReader.java:311)
	at org.eclipse.jgit.internal.storage.file.FileRepository.<init>(FileRepository.java:161)
	at org.eclipse.jgit.lib.BaseRepositoryBuilder.build(BaseRepositoryBuilder.java:625)
	at org.eclipse.jgit.api.Git.open(Git.java:93)
	at org.eclipse.jgit.api.Git.open(Git.java:73)
	at pl.allegro.tech.build.axion.release.infrastructure.git.GitRepository.<init>(GitRepository.java:70)
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:67)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:484)
	at org.codehaus.groovy.reflection.CachedConstructor.invoke(CachedConstructor.java:72)
	at org.codehaus.groovy.runtime.callsite.ConstructorSite$ConstructorSiteNoUnwrapNoCoerce.callConstructor(ConstructorSite.java:105)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallConstructor(CallSiteArray.java:59)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:263)
	at org.gradle.internal.classpath.intercept.CallInterceptorsSet$DecoratingCallSite.access$401(CallInterceptorsSet.java:128)
	at org.gradle.internal.classpath.intercept.CallInterceptorsSet$DecoratingCallSite$4.callOriginal(CallInterceptorsSet.java:180)
	at org.gradle.internal.classpath.intercept.CallInterceptorsSet$1.doIntercept(CallInterceptorsSet.java:54)
	at org.gradle.internal.classpath.intercept.CallInterceptorsSet$DecoratingCallSite.callConstructor(CallInterceptorsSet.java:177)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callConstructor(AbstractCallSite.java:277)
	at pl.allegro.tech.build.axion.release.infrastructure.di.ScmRepositoryFactory.create(ScmRepositoryFactory.groovy:23)
	at pl.allegro.tech.build.axion.release.infrastructure.di.ScmRepositoryFactory$create.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:139)
	at pl.allegro.tech.build.axion.release.infrastructure.di.VersionResolutionContext.create(VersionResolutionContext.groovy:43)
	at pl.allegro.tech.build.axion.release.infrastructure.di.VersionResolutionContext$create.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:148)
	at pl.allegro.tech.build.axion.release.infrastructure.di.VersionSupplier.resolve(VersionSupplier.groovy:10)
	at pl.allegro.tech.build.axion.release.infrastructure.di.VersionSupplier$resolve.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:148)
	at pl.allegro.tech.build.axion.release.infrastructure.di.MemoizedVersionSupplier.resolve(VersionSupplier.groovy:25)
	at pl.allegro.tech.build.axion.release.infrastructure.di.MemoizedVersionSupplier$resolve.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:148)
	at pl.allegro.tech.build.axion.release.domain.VersionConfig$_versionProvider_closure4.doCall(VersionConfig.groovy:221)
	at pl.allegro.tech.build.axion.release.domain.VersionConfig$_versionProvider_closure4.doCall(VersionConfig.groovy)
```

</details>


The stacktrace points at `pl.allegro.tech.build.axion.release.infrastructure.git.GitRepository.<init>(GitRepository.java:70)` as the offender.
Related gradle docs: https://docs.gradle.org/8.1/userguide/configuration_cache.html#config_cache:requirements:external_processes

The interesting part is I don't observe the error at every run, it pops up at occasionally (a guess: when gradle tries to create new configuration cache entry?)

--- a/src/main/java/pl/allegro/tech/build/axion/release/infrastructure/git/GitRepository.java
+++ b/src/main/java/pl/allegro/tech/build/axion/release/infrastructure/git/GitRepository.java
@@ -27,6 +27,7 @@ import org.eclipse.jgit.transport.RemoteRefUpdate;
 import org.eclipse.jgit.transport.TagOpt;
 import org.eclipse.jgit.transport.URIish;
 import org.eclipse.jgit.treewalk.filter.PathFilter;
+import org.eclipse.jgit.util.SystemReader;
 import org.eclipse.jgit.util.io.DisabledOutputStream;
 import org.gradle.api.logging.Logger;
 import org.gradle.api.logging.Logging;
@@ -66,6 +67,8 @@ public class GitRepository implements ScmRepository {
     private final ScmProperties properties;
 
     public GitRepository(ScmProperties properties) {
+        SystemReader.setInstance(new SystemReaderWithoutSystemConfig());
+
         try {
             this.repositoryDir = properties.getDirectory();
             this.jgitRepository = Git.open(repositoryDir);

--- a/src/main/java/pl/allegro/tech/build/axion/release/infrastructure/git/SystemReaderWithoutSystemConfig.java
+++ b/src/main/java/pl/allegro/tech/build/axion/release/infrastructure/git/SystemReaderWithoutSystemConfig.java
@@ -1,0 +1,73 @@
+package pl.allegro.tech.build.axion.release.infrastructure.git;
+
+import org.eclipse.jgit.lib.Config;
+import org.eclipse.jgit.storage.file.FileBasedConfig;
+import org.eclipse.jgit.util.FS;
+import org.eclipse.jgit.util.SystemReader;
+
+public class SystemReaderWithoutSystemConfig extends SystemReader
+{
+    private static final SystemReader DefaultSystemReader = SystemReader.getInstance();
+
+    @Override
+    public String getenv(String variable)
+    {
+        return DefaultSystemReader.getenv(variable);
+    }
+
+    @Override
+    public String getHostname()
+    {
+        return DefaultSystemReader.getHostname();
+    }
+
+    @Override
+    public String getProperty(String key)
+    {
+        return DefaultSystemReader.getProperty(key);
+    }
+
+    @Override
+    public long getCurrentTime()
+    {
+        return DefaultSystemReader.getCurrentTime();
+    }
+
+    @Override
+    public int getTimezone(long when)
+    {
+        return DefaultSystemReader.getTimezone(when);
+    }
+
+    @Override
+    public FileBasedConfig openUserConfig(Config parent, FS fs)
+    {
+        return DefaultSystemReader.openUserConfig(parent, fs);
+    }
+
+    @Override
+    public FileBasedConfig openJGitConfig(Config parent, FS fs) {
+        return DefaultSystemReader.openJGitConfig(parent, fs);
+    }
+
+    // Return an empty system configuration to prevent JGit from accessing it
+    // This resolves issues with Gradle being unable to save configuration cache
+    // Based on https://stackoverflow.com/a/59110721
+    @Override
+    public FileBasedConfig openSystemConfig(Config parent, FS fs)
+    {
+        return new FileBasedConfig(parent, null, fs)
+        {
+            @Override
+            public void load()
+            {
+            }
+
+            @Override
+            public boolean isOutdated()
+            {
+                return false;
+            }
+        };
+    }
+}

